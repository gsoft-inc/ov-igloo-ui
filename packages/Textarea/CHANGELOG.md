# @igloo-ui/textarea

## 0.2.4

### Patch Changes

- 876e9e1: Updated the autosize package for the textarea to fix the issue of the auto resizing not happening with preloaded text.

## 0.2.3

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.

## 0.2.2

### Patch Changes

- e876d94: Update textarea to use the same character count hook as the input component

## 0.2.1

### Patch Changes

- a075fd0: Fix the display of the negative character count when adding an emoji

## 0.2.0

### Minor Changes

- 7dbf543: Feat: added forward ref and autoFocus to textarea

## 0.1.5

### Patch Changes

- f47833b: Textarea - added disabled prop to the actual textarea

## 0.1.4

### Patch Changes

- 6ee9c45: Updated anything causing errors or warnings in the browser console

## 0.1.3

### Patch Changes

- 0640409: Update package autosize to 5.0.1
- 0640409: udpate classnames to 2.3.2

## 0.1.2

### Patch Changes

- d971143: All tests were updated to use testing library and not enzyme

## 0.1.1

### Patch Changes

- b8729ca: Feat: Initial release of the form group component

## 0.1.0

### Minor Changes

- 5e2bd6e: Feat: Initial release of the textarea component
