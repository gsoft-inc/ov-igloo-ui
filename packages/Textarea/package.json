{
  "name": "@igloo-ui/textarea",
  "version": "0.2.1",
  "main": "dist/Textarea.js",
  "module": "dist/Textarea.js",
  "types": "dist/Textarea.d.ts",
  "styles": "dist/textarea.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "@react-aria/utils": "^3.15.0",
    "autosize": "^5.0.1",
    "classnames": "^2.3.2",
    "graphemer": "^1.4.0"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ],
  "devDependencies": {
    "@types/autosize": "^4.0.1"
  }
}
