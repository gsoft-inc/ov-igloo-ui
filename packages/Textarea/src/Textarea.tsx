import * as React from 'react';
import cx from 'classnames';
import autosize from 'autosize';

import './textarea.scss';

export interface TextareaProps extends React.ComponentProps<'textarea'> {
  /** True if the textarea should allow new lines with Enter */
  allowNewline?: boolean;
  /** Add class names to the surrounding DOM container. */
  className?: string;
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** True if the textarea has an error. */
  error?: boolean;
  /** True if the textarea should resize
   * automatically when a multiline value is set. */
  isAutoResize?: boolean;
  /** The maximum number of characters allowed. Required if
   *  "showCharactersIndicator" is enabled. */
  maxLength?: number;
  /** Function called when the value changes. */
  onChange?: (e: React.ChangeEvent<HTMLTextAreaElement>) => void;
  /** True to display the number of remaining allowed characters
   *  in the bottom right of the text area. Requires a valid
   *  value in the "maxLength" prop. */
  showCharactersIndicator?: boolean;
  /** Specifies the value inside the input. */
  value?: string;
}

const Textarea: React.FunctionComponent<TextareaProps> = ({
  allowNewline = true,
  className,
  dataTest,
  error = false,
  isAutoResize = false,
  maxLength,
  onChange,
  showCharactersIndicator = false,
  value,
  ...props
}) => {
  const textareaRef = React.useRef<HTMLTextAreaElement | null>(null);
  const [currentCharLength, setCurrentCharLength] = React.useState(
    value?.length ?? 0
  );
  const [currentValue, setCurrentValue] = React.useState(value ?? '');
  const textareaMaxLength = maxLength ?? 0;
  const displayCharIndicator = showCharactersIndicator && textareaMaxLength > 0;

  const getClasses = (prefix = 'ids-textarea'): string => {
    const classes = cx(prefix, {
      [`${prefix}--error`]: error,
      [`${prefix}--has-char-count`]: displayCharIndicator,
      [`${prefix}--disabled`]: props.disabled,
    });

    return classes;
  };

  const handleOnChange = (
    event: React.ChangeEvent<HTMLTextAreaElement>
  ): void => {
    setCurrentValue(event.target.value);
    if (onChange) {
      onChange(event);
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>): void => {
    if (e.key === 'Enter') {
      e.stopPropagation();

      if (!allowNewline) {
        e.preventDefault();
      }
    }
  };

  React.useEffect(() => {
    if (isAutoResize && textareaRef && textareaRef.current) {
      autosize(textareaRef.current);
      autosize.update(textareaRef.current);
    }
  }, [textareaRef, isAutoResize]);

  React.useEffect(() => {
    setCurrentCharLength(currentValue?.length ?? 0);
  }, [currentValue]);

  return (
    <div className={`${getClasses()} ${className}`} data-test={dataTest}>
      <textarea
        ref={textareaRef}
        className={getClasses('ids-textarea__field')}
        maxLength={maxLength}
        onChange={handleOnChange}
        onKeyDown={handleKeyDown}
        value={value}
        {...props}
      />

      {displayCharIndicator && (
        <div className="ids-textarea__char-indicator">
          {textareaMaxLength - currentCharLength}
        </div>
      )}
    </div>
  );
};

export default Textarea;
