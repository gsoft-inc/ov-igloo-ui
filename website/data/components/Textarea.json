[
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/Textarea/src/Textarea.tsx",
      "description": "",
      "displayName": "Textarea",
      "methods": [],
      "props": {
        "allowNewline": {
          "defaultValue": null,
          "description": "True if the textarea should allow new lines with Enter",
          "name": "allowNewline",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "className": {
          "defaultValue": null,
          "description": "Add class names to the surrounding DOM container.",
          "name": "className",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "dataTest": {
          "defaultValue": null,
          "description": "Add a data-test tag for automated tests",
          "name": "dataTest",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "disabled": {
          "defaultValue": null,
          "description": "True if the textarea should be disabled.",
          "name": "disabled",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "error": {
          "defaultValue": null,
          "description": "True if the textarea has an error.",
          "name": "error",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "isAutoResize": {
          "defaultValue": null,
          "description": "True if the textarea should resize\nautomatically when a multiline value is set.",
          "name": "isAutoResize",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "maxLength": {
          "defaultValue": null,
          "description": "The maximum number of characters allowed. Required if\n\"showCharactersIndicator\" is enabled.",
          "name": "maxLength",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "number" }
        },
        "onChange": {
          "defaultValue": null,
          "description": "Function called when the value changes.",
          "name": "onChange",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "(e: ChangeEvent<HTMLTextAreaElement>) => void" }
        },
        "showCharactersIndicator": {
          "defaultValue": null,
          "description": "True to display the number of remaining allowed characters\nin the bottom right of the text area. Requires a valid\nvalue in the \"maxLength\" prop.",
          "name": "showCharactersIndicator",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "value": {
          "defaultValue": null,
          "description": "Specifies the value inside the input.",
          "name": "value",
          "parent": {
            "fileName": "components/packages/Textarea/src/Textarea.tsx",
            "name": "TextareaProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Textarea/src/Textarea.tsx",
              "name": "TextareaProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        }
      }
    }
  ]
]
