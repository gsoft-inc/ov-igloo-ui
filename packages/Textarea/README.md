# Textarea

The textarea component defines a multi-line text input control.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/textarea` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/textarea
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/textarea
```

## Usage

Then to use the component in your code just import it!

```jsx
import Textarea from '@igloo-ui/textarea';

<Textarea
  placeholder="The character count will count down to show how many characters are left. MaxLength is required"
  maxLength={1000}
  showCharactersIndicator
/>;
```
