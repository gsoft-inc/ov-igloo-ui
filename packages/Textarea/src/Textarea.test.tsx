/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';

import Textarea from './Textarea';

const setUp = (props = {}) => {
  const component = shallow(<Textarea dataTest="ids-textarea" {...props} />);
  return component;
};

describe('Textarea', () => {
  test('It should render without errors', () => {
    const component = setUp();
    const wrapper = component.find('.ids-textarea');
    expect(wrapper.length).toBe(1);
  });

  test('It should render a snapshot', () => {
    const component = setUp();
    expect(component).toMatchSnapshot();
  });

  test('It should render disabled', (): void => {
    const component = setUp({ disabled: true });
    const wrapper = component.find('.ids-textarea--disabled');
    expect(wrapper.length).toBe(1);
  });

  test('It should render an error look', (): void => {
    const component = setUp({ error: true });
    const wrapper = component.find('.ids-textarea--error');
    expect(wrapper.length).toBe(1);
  });

  test('It should render a character count', (): void => {
    const component = setUp({ showCharactersIndicator: true, maxLength: 100 });
    const wrapper = component.find('.ids-textarea__char-indicator');
    expect(wrapper.length).toBe(1);
  });

  test('It should not render a character count without max length', (): void => {
    const component = setUp({ showCharactersIndicator: true });
    const wrapper = component.find('.ids-textarea__char-indicator');
    expect(wrapper.length).toBe(0);
  });

  test('It should call the onChange function', () => {
    let changed = false;
    const event = {
      target: { name: 'textareaname', value: 'value of the textarea' },
    };
    const props = {
      onChange: () => {
        changed = true;
      },
    };
    const component = setUp(props);

    component.find('.ids-textarea__field').simulate('change', event);
    expect(changed).toBeTruthy();
  });
});
