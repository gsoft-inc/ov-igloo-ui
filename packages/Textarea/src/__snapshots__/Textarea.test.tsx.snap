// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea It should render a snapshot 1`] = `ShallowWrapper {}`;
